File: src/DataScience/Octave/bridge.m
function bridge()
  tv = make_device("TV");
  radio = make_device("Radio");

  fprintf("basic-tv=%s\n", activate_basic(tv));
  fprintf("basic-radio=%s\n", activate_basic(radio));
  fprintf("mute-tv=%s\n", activate_mute(tv));
  fprintf("mute-radio=%s\n", activate_mute(radio));
end

function device = make_device(name)
  device.power_on = @() [name ":on"];
  device.mute = @() [name ":muted"];
end

function result = activate_basic(device)
  result = device.power_on();
end

function result = activate_mute(device)
  result = device.mute();
end

File: src/Niche/Nim/bridge.nim
type
  Action = proc (): string
  Device = object
    powerOn: Action
    mute: Action

proc makeDevice(name: string): Device =
  result.powerOn = proc (): string = name & ":on"
  result.mute = proc (): string = name & ":muted"

proc activateBasic(device: Device): string = device.powerOn()
proc activateMute(device: Device): string = device.mute()

let tv = makeDevice("TV")
let radio = makeDevice("Radio")

echo "basic-tv=" & activateBasic(tv)
echo "basic-radio=" & activateBasic(radio)
echo "mute-tv=" & activateMute(tv)
echo "mute-radio=" & activateMute(radio)
